### PR TITLE
Add a metric for tracking TODO's in source code. Closes #5630.

### LINKED ISSUE #5630: Add new metric which measures TODO's in source code
**Describe the solution you'd like**
Add support for (with a new metric which measures) TODO's in code to automatically support/measure: NPR5325 6.4 "Demonstrate that traces of unfinished work are absent or limited in number or size. An excess of work in progress (to do's, temporary facilities) may mean that construction has not yet been completed or that the system contains hidden defects." 

In Dutch: "Toon aan dat sporen van niet afgeronde werkzaamheden afwezig of beperkt in aantal of in omvang zijn. Een teveel aan onderhanden werk (to do's, tijdelijke voorzieningen) kan betekenen dat de bouw nog niet is afgerond of dat het systeem verborgen gebreken bevat." 

Each language in SonarQube has a rule: "Track uses of "TODO" tags". 
Query: https://sonar.instance.country/coding_rules?q=Track+uses+of+%22TODO%22+tags

Or use the suffix: :S1135 e.g. csharpsquid:S1135 or swift:S1135

### PR DESCRIPTION
**Pull request by fniessink**

### CODE REVIEW

File: components/shared_code/src/shared_data_model/meta/unit.py [hunk 1]
```diff
@@ -29,6 +29,7 @@ class Unit(StrEnum):
     SUPPRESSED_VIOLATIONS = "suppressed violations"
     TEST_CASES = "test cases"
     TESTS = "tests"
+    TODOS = "to-do's"
```
Comment (wkoot) [inline components/shared_code/src/shared_data_model/meta/unit.py]: Is there a better term somehow? Plural should be "to-dos" according to both https://www.merriam-webster.com/dictionary/to-do and https://www.collinsdictionary.com/dictionary/english/to-do .. But however not listed in the Cambridge dictionary

File: components/shared_code/src/shared_data_model/metrics.py [hunk 2]
```diff
@@ -609,6 +609,20 @@
         near_target="14",
         sources=["calendar"],
     ),
+    "todos": Metric(
+        name="To-do's",
+        description="The number of to-do's in source code.",
+        rationale="Code should not contain traces of unfinished work. An excess of work in progress (to-do's, fixme's) "
```
Comment (wkoot) [inline components/shared_code/src/shared_data_model/metrics.py]: I wouldn't call it an "excess of work in progress", that would be more fitting for other metrics. Is it not directly technical debt, or some other form of clutter?

File: components/shared_code/src/shared_data_model/metrics.py [hunk 3]
```diff
@@ -609,6 +609,20 @@
...
+        "may mean that construction has not yet been completed or that the system contains hidden defects.",
+        rationale_urls=[
+            "https://rules.sonarsource.com/python/RSPEC-1135/",
+            "https://cwe.mitre.org/data/definitions/546",
+        ],
+        unit=Unit.TODOS,
+        near_target="100",
```
Comment (wkoot) [inline components/shared_code/src/shared_data_model/metrics.py]: Feels pretty high

File: components/shared_code/src/shared_data_model/sources/sonarqube.py [hunk 4]
```diff
@@ -343,6 +392,7 @@ def violation_entity_attributes(
             name="violation",
             attributes=violation_entity_attributes(include_resolution=True, include_rationale=True),
         ),
+        "todos": VIOLATION_ENTITY,
```
Comment (wkoot) [inline components/shared_code/src/shared_data_model/sources/sonarqube.py]: But this is `name="violations"`? Weird btw that the one on the next line isn't even using `VIOLATION_ENTITY`

**Comment by denniebouman** [comment]
LGTM!

**Review by denniebouman — COMMENTED** [review]
LGTM!

File: components/shared_code/src/shared_data_model/meta/unit.py (hunk 1 above)
Comment (fniessink) [reply components/shared_code/src/shared_data_model/meta/unit.py]: I've googled a bit for a more generic term that encapsulates both to-do and fixme. I've seen "task tag" (used in some IDE's apparently) and "codetag" (in a rejected Python PEP), but those are not as self-explanatory as to-do. Maybe "to-do and fix-me comments" is better?

File: components/shared_code/src/shared_data_model/metrics.py (hunk 2 above)
Comment (fniessink) [reply components/shared_code/src/shared_data_model/metrics.py]: I'll reword.

File: components/shared_code/src/shared_data_model/metrics.py (hunk 3 above)
Comment (fniessink) [reply components/shared_code/src/shared_data_model/metrics.py]: Agreed. Set to 50.

File: components/shared_code/src/shared_data_model/sources/sonarqube.py (hunk 4 above)
Comment (fniessink) [reply components/shared_code/src/shared_data_model/sources/sonarqube.py]: SonarQube is reporting rule violations, hence the measurement entities are called violations. Fixed the "next line".

File: components/shared_code/src/shared_data_model/meta/unit.py (hunk 1 above)
Comment (fniessink) [reply components/shared_code/src/shared_data_model/meta/unit.py]: I'm going to use "todo and fixme comments" as that also prevents the ugly plural "to-dos" or "to-do's". I know todo should be spelled to-do, but that's not how it is typically written in code comments, so...

Comment (wkoot) [reply components/shared_code/src/shared_data_model/meta/unit.py]: Sounds like a great solution. It's very fitting to use "todo" next to "fixme" anyway

**Pull request merged**